examples/seismic/source.py

```diff
@@ -8,53 +10,100 @@
-    :param npoint: (Optional) Number of sparse points represented by this source
-    :param time_order: (Optional) Time discretization order (defaults to 2)
+    :param coordinates: Point coordinates for this source.
+    :param time_range: (Optional) Tuple of time origin, time interval and
+                       number of timesteps, (t0, dt, ntime). Can only set
+                       one of time_range and time.
+    :param time: (Optional) Time series. Can only set one of time_range and time.
+    :param data: (Optional) Data values to initialise point data.
+    :param npoint: (Optional) Number of sparse points represented by this source.
+    :param time_order: (Optional) Time discretization order (defaults to 2).
     :param dimension: :(Optional) class:`Dimension` object for
-                       representing the number of points in this source
+                       representing the number of points in this source.
 
     Note, either the dimensions `ntime` and `npoint` or the fully
     initialised `data` array need to be provided.
     """
 
-    def __new__(cls, name, grid, ntime=None, npoint=None, data=None,
-                coordinates=None, **kwargs):
+    def __new__(self, name, grid, time_range=None, time=None, npoint=None,
+                data=None, coordinates=None, **kwargs):
         p_dim = kwargs.get('dimension', Dimension(name='p_%s' % name))
         time_order = kwargs.get('time_order', 2)
         npoint = npoint or coordinates.shape[0]
-        if data is None:
-            if ntime is None:
-                error('Either data or ntime are required to'
-                      'initialise source/receiver objects')
+
+        if time_range is None and time is None:
+            error("Either time_range or time must be defined.")
+
+        if time_range is None:
+            self._time_range = (time[0], time[1]-time[0], time.size)
         else:
-            ntime = ntime or data.shape[0]
+            self._time_range = time_range
+        nt = self._time_range[2]
 
         # Create the underlying SparseTimeFunction object
-        obj = SparseTimeFunction.__new__(cls, name=name, grid=grid,
+        obj = SparseTimeFunction.__new__(self, name=name, grid=grid,
                                          dimensions=[grid.time_dim, p_dim],
-                                         npoint=npoint, nt=ntime, time_order=time_order,
+                                         npoint=npoint, nt=nt,
+                                         time_order=time_order,
                                          coordinates=coordinates, **kwargs)
 
         # If provided, copy initial data into the allocated buffer
         if data is not None:
             obj.data[:] = data
+
         return obj
 
     def __init__(self, *args, **kwargs):
         if not self._cached():
             super(PointSource, self).__init__(*args, **kwargs)
 
+    @property
+    def t_start(self):
```

Comment:
is this expected to be called from the outside/user? if not, make it private

Reply:
It can be called outside in cases such as when we need to clone the receiver to store the data misfit. I think there is an argument for making it all private and adding a member function to do the cloning. Thoughts?

Reply:
agreed. 

Probably unrelated to this PR: Is cloning an operation that we expect to be needed on other kind of objects? if so, we should provide a superclass capturing this "clonable" property

Reply:
For that moment I made most of these private as you suggested.